File: surf-cloud-api/src/main/java/dev/slne/surf/cloud/api/SurfCloudApplication.java
package dev.slne.surf.cloud.api;

import dev.slne.surf.cloud.api.config.JpaConfig;
import dev.slne.surf.cloud.api.config.RedisConfig;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.autoconfigure.data.redis.RedisRepositoriesAutoConfiguration;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.core.annotation.AliasFor;
import org.springframework.scheduling.annotation.EnableAsync;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.transaction.annotation.EnableTransactionManagement;

@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@SpringBootApplication(exclude = {DataSourceAutoConfiguration.class, RedisRepositoriesAutoConfiguration.class})
@EnableScheduling
@EnableAsync
@EnableTransactionManagement
@EnableAspectJAutoProxy
@JpaConfig
@RedisConfig
@EntityScan
//@EnableRedisRepositories
public @interface SurfCloudApplication {

  @AliasFor(annotation = SpringBootApplication.class, attribute = "scanBasePackages")
  String[] basePackages() default {};

//  @AliasFor(annotation = JpaConfig.class, attribute = "basePackages")
  String[] jpaBasePackages() default {};

//  @AliasFor(annotation = RedisConfig.class, attribute = "basePackages")
  String[] redisBasePackages() default {};
}

File: surf-cloud-api/surf-cloud-api-common/src/main/kotlin/dev/slne/surf/cloud/api/common/SurfCloudApplication.kt
package dev.slne.surf.cloud.api.common

import org.springframework.boot.autoconfigure.AutoConfigurationExcludeFilter
import org.springframework.boot.autoconfigure.AutoConfigurationPackage
import org.springframework.boot.autoconfigure.EnableAutoConfiguration
import org.springframework.boot.autoconfigure.domain.EntityScan
import org.springframework.boot.context.TypeExcludeFilter
import org.springframework.cache.annotation.EnableCaching
import org.springframework.context.annotation.AdviceMode
import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.FilterType
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.scheduling.annotation.EnableScheduling
import java.lang.annotation.Inherited


/**
 * Annotation for defining Surf Cloud applications with specific configurations.
 *
 * Combines multiple Spring features such as scheduling, async processing,
 * transaction management, and entity scanning, with options for JPA and Redis base packages.
 */
@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
@EnableScheduling
@EnableAsync(mode = AdviceMode.ASPECTJ)
//@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
@EntityScan
@EnableCaching(mode = AdviceMode.ASPECTJ)
@AutoConfigurationPackage
@Inherited
@EnableAutoConfiguration
@ComponentScan(
    excludeFilters = [ComponentScan.Filter(
        type = FilterType.CUSTOM,
        classes = [TypeExcludeFilter::class]
    ), ComponentScan.Filter(
        type = FilterType.CUSTOM,
        classes = [AutoConfigurationExcludeFilter::class]
    )]
)
annotation class SurfCloudApplication
